OMCompiler/Compiler/Util/ContainerImage.mo

```diff
@@ -0,0 +1,427 @@
+      Error.addCompilerNotification(System.readFile(cosignLogFile) + "\n");
+      System.removeFile(cosignLogFile);
+      // TODO: Fail if failing to verify!
+      // fail();
+      return;
+    end if;
+
+    System.removeFile(cosignLogFile);
+  end assertSignature;
+
+  function toString
+    "Return container image as string in format
+     [[HOST[:PORT]/]NAMESPACE/]REPOSITORY[:TAG]."
+    input ContainerImage image;
+    output String imageString = "";
+  algorithm
+    imageString := hostToString(image);
+    if not imageString == "" then
+      imageString := imageString + "/";
+    end if;
+
+    imageString := imageString + nameToString(image);
+
+    if isSome(image.tag) then
+      imageString := imageString + ":" + Util.getOption(image.tag);
+    end if;
+  end toString;
+
+  function nameToString
+    "Return name [NAMESPACE/]REPOSITORY as String."
+    input ContainerImage image;
+    output String name = "";
+  algorithm
+    name := match image
+      local
+        String namespace_str;
+        String repository;
+      case CONTAINER_IMAGE(namespace = SOME(namespace_str), repository = repository)
+        then(namespace_str + "/" + repository);
+      case CONTAINER_IMAGE(namespace = NONE(), repository = repository)
+        then(repository);
+      else
+      algorithm
+        Error.addCompilerError("Failed to get name of image reference.");
+        then fail();
+    end match;
+  end nameToString;
+
+protected
+  // Helper functions for parsing
+  function parseContainerHostPort
+    "Parse container host and port string:
+        HOST[:PORT]"
+    input String host_and_port;
+    output String host;
+    output Option<String> port := NONE();
+  algorithm
+    host := match Util.stringSplitAtChar(host_and_port, ":")
+    local
+      String host_str;
+      String port_str;
+      case {host_str, port_str}
+      algorithm
+        port := SOME(port_str);
+      then(host_str);
+      case {host_str} then(host_str);
+      else
+      algorithm
+        Error.addCompilerError("Failed to parse container host '" + host_and_port + "'.");
+        then fail();
+    end match;
+  end parseContainerHostPort;
+
+  function parseContainerRepository
+    "Parse container image repository string:
+        REPOSITORY[:TAG]"
+    input String repositoryString;
+    output String repository;
+    output Option<String> tag;
+  algorithm
+    _ := match Util.stringSplitAtChar(repositoryString, ":")
+    local
+      String repository_str;
+      String tag_str;
+    case {repository_str, tag_str}
+    algorithm
+      repository := repository_str;
+      tag := SOME(tag_str);
+      then();
+    case {repository_str}
+    algorithm
+      repository := repository_str;
+      tag := NONE();
+      then();
+    else
+    algorithm
+      Error.addCompilerError("Failed to parse container image name '" + repositoryString + "'.");
+      then fail();
+    end match;
+  end parseContainerRepository;
+
+  // Helper functions for toString
```

Comment:
Helper functions could be put in the protected section if you like to hide them from the outside.
```suggestion
protected
  // Helper functions for toString
```

Reply:
All these functions are already in the protected sub-section.
Or would you add protected to all protected functions?

Reply:
Ah no it's fine, I overlooked that.